### ISSUE TITLE
How do you dynamically load an Icon?

### ISSUE THREAD

**Issue by ElixirMike** [Templarian/MaterialDesign-React#34]
So, in my app, I want to provide an Icon picker...let the user choose their icon. Once saved, I would use it on other pages in the app.

How do I dynamically import this specific icon on those pages?

**Comment by Templarian** [comment]
Yeah, we need to document this process and how it should work. It unfortunately requires a server-side component if one wants to use all 5k icons or it can become a performance issue for pages with lots of icons.

Do you want to provide all ~5k icons in the icon picker or a subset? We/community need to make some demos of both examples.

**Comment by ElixirMike** [comment]
I was just looking to create a subset...we've picked out about 800 icons to use. For now, what I did was create a custom object with array of icons, including a tag, name of icon and path.

Like this where I insert the 800 icons.
var myicons = [{tag:'phones', name='mdiphone', path: "....full svg path", {}, {}, etc ]

It's lazy loaded on icon picker page..but when a user selects the icon, I simply store the svg path in db/cache. 

Then, in the rest of my application, I can simply use the svgpath and display, so it's very fast.

It would be super helpeful to have someway of picking icons you want and then generating this type of object which can be used to display them on icon selection page.

**Comment by AmirHosseinKarimi** [comment]
I am trying to use dynamic import from [this article](https://kamranicus.com/posts/2017-09-02-dynamic-import-material-icons-react)
But in this article only can import whole file, refer to @mdi/js/mdi.js all icons are stored to this single file.
Is there anyway to dynamically import specific icon from this file?

**Comment by Templarian** [comment]
@AmirHosseinKarimi You could use the `@mdi/svg` package and fetch the individual SVG's by their icon name in a wrapper `<LazyIcon name={iconName}>`. Since `<Icon>` takes any path data, you could regex for the path data out of the SVG request and put it into `<Icon>`.

```ts
const data = body.match(/d="[^"]+"/)[0];
```

Obviously write a light cache layer into your component so you don't make unnecessary requests for the path data.

**Comment by Templarian** [comment]
The `@mdi/js` package is only really aimed at compile time for tree-shaking. The `@mdi/svg` also contains the `meta.json` if you need to generate a list of icons (recommend to trim down this file as it could be heavy for use on the web).

**Comment by AmirHosseinKarimi** [comment]
Sorry I did not understand

**Comment by Templarian** [comment]
@AmirHosseinKarimi https://github.com/Templarian/MaterialDesign-SVG contains individual SVG files you can deploy. These could be requested by their name in a wrapping component. You only need the SVG path data aka `<path d="this content"/>` so that regex will get that (made a typo and edited it above with the correct regex).

**Comment by AmirHosseinKarimi** [comment]
So I have created a [npm package](https://www.npmjs.com/package/mdi-js) for dynamically load an icon.
The description is clear about how this package working.
Its a dependency of a package which I will create for React that automatically load the icon.

Currently if you need a dynamically load icon, you can create the following component:
```javascript
import React from "react";
import { Icon as MDIcon } from "@mdi/react";

class Icon extends React.Component {
  render() {
    let icon = require(`mdi-js/icons/${this.props.icon}`).default;
    if (!icon) {
      throw Error(`Could not find mdi-js/icons/${icon}`);
    }
    const iconName = this.props.icon.substr(3);
    return <MDIcon path={icon} />;
  }
}

export default Icon;
```
and use like this:
```javascript
import Icon from "./Icon";
...
<Icon icon="mdiAccount" />
```

If this package is helped you please give a star: ![GitHub stars](https://img.shields.io/github/stars/AmirHosseinKarimi/MaterialDesign-JS?style=social)

Also if you like to contribute, just look at the [MaterialDesign-JS](https://github.com/AmirHosseinKarimi/MaterialDesign-JS) repository.

**Comment by Templarian** [comment]
We recommend using a peer dependency for `@mdi/js`. Referencing a specific version usually doesn't work out in the long run. Learn more:

- https://dev.materialdesignicons.com/contribute/third-party
- https://dev.materialdesignicons.com/guide/nodejs-scripting

Yours will be more lightweight than the `@mdi/svg` approach, but similar idea loading individual glyphs.

We will work to open source the IndexedDB solution as it will be more ideal performance wise. Also makes it easier in cases where one may need to search all icons in a list.

**Comment by AmirHosseinKarimi** [comment]
Good, I create this package because I need dynamic load icons in my current project.
So always there is an better idea. I am waiting for IndexedDB solution to try it.

**Comment by AmirHosseinKarimi** [comment]
Finally I have find a way which can use dynamic load MDI and also do not decrease the performance of app in development and runtime.
Just look at this gist: https://gist.github.com/AmirHosseinKarimi/801931665a0845ccc25ed486431abf72

**Comment by craigrileyuk** [comment]
> https://gist.github.com/AmirHosseinKarimi/801931665a0845ccc25ed486431abf72

You're requiring the entire @mdi/js library here, so any development or production build of the app would have to include the whole package (around 2.3MB).

**Comment by AmirHosseinKarimi** [comment]
@craigrileyuk Yea you're right. It's a little bit old. But I think the dynamic load example in the [MaterialDesign-JS](https://github.com/AmirHosseinKarimi/MaterialDesign-JS)  repository readme is correct and working.

**Comment by ironAngel2000** [comment]
maybe a little late but this works for me fine:

```
import React, { useEffect, useState } from "react";
import { Icon as MDIcon } from "@mdi/react";

type props = {
    icon: string,
    size: number,
}

const LazyIcon = (props: props) => {

    const [iconPath, setIconPath] = useState<string>('');

    useEffect(() => {
        import('@mdi/js').then((icons:any) => {
            if (icons[props.icon]) {
                const icon = icons[props.icon];
                setIconPath(icon.toString());
            }
        });
    }, [props.icon]);

    return (<>{iconPath !== '' ?
        <MDIcon path={iconPath} size={props.size}></MDIcon>
        : null
    }</>);

}

export { LazyIcon };
```
